geps/gep-713.md

```diff
@@ -38,24 +58,162 @@ portable.
+of those objects. Sometimes, this is because we can’t change the spec of the object
+to hold the information we need ( ReferenceGrant, from GEP-709, affecting Secrets
+and Services is an example, as is Direct Policy Attachment), and sometimes it’s
+because we want the behavior change to flow across multiple objects
+(this is what Hierarchical Policy Attachment is for).
+
+To put this another way, sometimes we need ways to be able to affect how an object
+is interpreted in the API, without representing the description of those effects
+inside the spec of the object.
+
+This document describes the ways we design objects to meet these two use cases,
+and why you might choose one or the other.
+
+We use the term “metaresource” to describe the class of objects that _only_ augment
+the behavior of another Kubernetes object, regardless of what they are targeting.
+
+“Meta” here is used in its Greek sense of “more comprehensive”
+or “transcending”, and “resource” rather than “object” because “metaresource”
+is more pronounceable than “metaobject”. Additionally, a single word is better
+than a phrase like “wrapper object” or “wrapper resource” overall, although both
+of those terms are effectively synonymous with “metaresource”.
+
+A "Policy Attachment" is a metaresource that affects the fields in existing objects
+(like Gateway or Routes), or influences the configuration that's generated in an
+underlying data plane.
+
+"Direct Policy Attachment" is when a Policy object references a single object _only_,
+and only modifies the fields of or the configuration associated with that object.
+
+"Hierarchical Policy Attachment" is when a Policy object references a single object
+_and any child objects of that object_ (according to some defined hierarchy), and
+modifies fields of the child objects, or configuration associated with the child
+objects.
+
+### Direct Policy Attachment
+
+A Direct Policy Attachment is tightly bound to one or more instances of a particular
+Kind within a single namespace, and only modifies the behavior of the object or
+objects that match its binding.
+
+As an example, one use case that Gateway API currently does not support is how
+to configure details of the TLS required to connect to a backend (in other words,
+if the process running inside the backend workload expects TLS, not that some
+automated infrastructure layer is provisioning TLS as in the Mesh case).
+
+A hypothetical TLSConnectionPolicy that targets a Service could be used for this,
+using the functionality of the Service as describing a set of endpoints. (It
+should be also noted this is not the only way to solve this problem, just an
+example to illustrate Direct Policy Attachment.)
+
+The TLSConnectionPolicy would look something like this:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: TLSConnectionPolicy
+metadata:
+  name: tlsport8443
+  namespace: foo
+spec:
+  targetRef: # This struct is defined as part of Gateway API
+    type: Direct
```

Comment:
I had expected Direct vs Hierarchical was a property of the CRD - whether it supports the more complex form or only the simplified one.

I don't really understand how a user would know to set either field. Can you show an example of the same policy type (TLSConnectionPolicy) in both modes and explain how they differ?

Reply:
I'm actually kind of using this field as a way for the CRD to indicate in a field inside the object which type it is - I'm thinking about how we can conceivably write tooling to generically parse an arbitrary Policy object, _from any implementation_, and we need a way for it to know what sort of object we're dealing with. At a minimum, I'd like an
arbitrary tool to have relatively-simple rule to allow it to guess if the Policy object is hierarchical or direct.

I guess I'm also trying to lean in to the statement `targetRef` is what makes a Policy object.

You're right that this doesn't make sense as a user-settable field, but neither does an annotation, I think, so I'm at a loss.

Maybe a better option would be to say "a Policy is an object with a `targetRef`, and a Hierarchical Policy is a Policy object with a `targetRef` and one or more of `defaults` or `overrides`." So, unless you use the standard way of doing Hierarchical Policy, then your policy must only affect that one object. I'll think about that some more, but hopefully this explains a little better what I was aiming at.

Reply:
Defaulting in the CRD could set it (and enforce it isn't set to anything else) perhaps?

Or the bit about defaults/overrides

Reply:
I though about this some more, and updated to make the defining feature of a hierarchical policy that it possesses a `targetRef` stanza, and then either a `defaults`, an `overrides`, or both. Not having at least one of `defaults` or `overrides` means that your Policy is a Direct Attached Policy.

Essentially, this reserves two special keys in the top level of every Policy CRD, `defauls`, and `overrides`, and uses their presence to imply behavior.

I don't think I can do much clearer than that without also being too clever. 😄

Reply:
I like this way (default and overrides vs just targetRef) of distinguishing direct vs Hierarchical so +1 from me

Reply:
+1 to how it's described [here](https://github.com/kubernetes-sigs/gateway-api/pull/1565/files#r1133422862) (and looks like the line this thread is attached to can be removed)

Reply:
This was resolved but I don't think the example was updated